requests-for-comments/2023-03-01-Temporal-RFC.md

```diff
@@ -0,0 +1,239 @@
+
+# Temporal
+
+Temporal is a workflow engine that provides a way for developers to build and manage distributed applications. It is designed to help developers focus on their business logic by abstracting away the details of failure modes, retry logic, and timeouts. Temporal allows developers to build workflows that describe complex processes as a set of simple, easy-to-understand steps. It provides a programmable interface that enables developers to write the code that defines their workflows and the logic that controls how they execute.
+
+Temporal is used by a number of companies, including Uber, DoorDash, Instacart, HashiCorp, and Coinbase. It has become a popular choice among developers because it provides a way to manage complex distributed systems without having to worry about the underlying infrastructure. Temporal allows developers to build workflows that are resilient, scalable, and fault-tolerant, which is essential for building applications that can handle large volumes of data and traffic.
+
+One of the key features of Temporal is its ability to provide visibility into the status of workflows. It provides a dashboard that allows developers to monitor the progress of their workflows and to see where errors are occurring. This helps developers to quickly identify and fix problems, which is essential for ensuring that workflows are running smoothly.
+
+Overall, Temporal is a powerful tool that enables developers to build complex distributed applications and workflows. It provides a way to manage the complexity of distributed systems and to build applications that are scalable, resilient, and fault-tolerant.
+
+[What is temporal?](https://docs.temporal.io/temporal)
+
+[How does temporal work?](https://temporal.io/how-temporal-works)
+
+### Why is it a good fit?
+
+Temporal is an especially good fit because it provides things that are genuinely hard to build and maintain ourselves:
+
+- Retry logic per Activity and Workflow
+- Timeout logic per activity and workflow
+- Exactly-once semantics (!!)
+- Go, [Python](https://docs.temporal.io/kb/python-sandbox-environment), and Typescript libraries that support deterministic code execution and retries (!!)
+- Composability of Activities and Workflows
+- Dashboard for observing Workflow status
+
+Finally, Temporal is a good fit because we are a small team. We should be focusing on the business of building our product. Nothing about what we are building that would be run on Temporal is novel enough to justify us to invest in a competing runtime.
+
+### Who uses it?
+
+- **Uber** (original developer - originally called [Cadence](https://github.com/uber/cadence))
+- **Coinbase**
+- **Doordash**
+- **Instacart**
+- **Hashicorp**
+- **Netflix**
+- **Snapchat**
+- **Stripe**
+- **Datadog**
+- **Box**
+- **Airbyte**
```

Comment:
what do they use it for?
There's a big difference in using an external scheduling system for something real time vs something that just needs to run daily/weekly/... reliably. Asking because if temporal has an outage we are not going to be able to do much about it if it's very core except schedule the tasks manually directly.

Reply:
Great question. At Uber it was used for very high frequency workflows. You've probably even been a customer who's order has gone through Cadence if you have ever ordered something on UberEats.

https://www.uber.com/blog/open-source-orchestration-tool-cadence-overview/
https://www.youtube.com/watch?v=-LRghQzfF8k

Though, keep in mind I am not advocating here that we process each event through this. The highest frequency would be webhook jobs.